runtime/runtime/src/bandwidth_scheduler/mod.rs

```diff
@@ -38,55 +48,89 @@ pub fn run_bandwidth_scheduler(
         Some(prev_state) => prev_state,
         None => {
             tracing::debug!(target: "runtime", "Bandwidth scheduler state not found - initializing");
-            BandwidthSchedulerState { mock_data: [0; 32] }
+            BandwidthSchedulerState {
+                link_allowances: Vec::new(),
+                sanity_check_hash: CryptoHash::default(),
+            }
         }
     };
 
-    let scheduler_state_hash: CryptoHash = hash(&borsh::to_vec(&scheduler_state).unwrap());
+    // Prepare the status info for each shard based on congestion info.
+    let mut shards_status: BTreeMap<ShardId, ShardStatus> = BTreeMap::new();
+    for (shard_id, extended_congestion_info) in apply_state.congestion_info.iter() {
+        let last_chunk_missing = extended_congestion_info.missed_chunks_count > 0;
+        let is_allowed_sender_shard =
+            *shard_id == ShardId::from(extended_congestion_info.congestion_info.allowed_shard());
+
+        let congestion_control = CongestionControl::new(
+            apply_state.config.congestion_control_config,
+            extended_congestion_info.congestion_info,
+            extended_congestion_info.missed_chunks_count,
+        );
+        let congestion_level = congestion_control.congestion_level();
+        let is_fully_congested = CongestionControl::is_fully_congested(congestion_level);
+
+        shards_status.insert(
+            *shard_id,
+            ShardStatus { last_chunk_missing, is_allowed_sender_shard, is_fully_congested },
+        );
+    }
 
+    // Prepare lists of sender and receiver shards.
+    // TODO(bandwidth_scheduler) - find a better way to get the sender and receiver shards.
+    // Maybe pass shard layout in ApplyState? That might also be needed for resharding.
+    // Taking all shards from the congestion info will be wrong for only one height during
+    // resharding, so it might be good enough for now, but it's not ideal.
     let mut all_shards = apply_state.congestion_info.all_shards();
```

Comment:
> Using congestion_info.all_shards() might not work here as this list only contains the shards that have some valid congestion info (i.e. there are some buffered receipts)

My understanding is that all shards have congestion info, even if they have zero delayed/postponed/buffered receipts and zero congestion, so it should work fine on real-world chains.

Changed to `ShardLayout` anyway, it's easier to think about.

Reply:
That's interesting, I can't think of any scenario where this would happen (apart from resharding). Bandwidth requests are propagated the same way, that'd mean that sometimes they won't reach other shards?

Reply:
We were considering removing the shard_id entries for which we don't have buffered receipts as one of the way to clear old resharding data. That's a protocol change, however it makes sense to have the rest of the design in such a way that we don't take dependency on having all shard_id values in there.

Reply:
Whatever happens with congestion info, AFAIU `BlockBandwidthRequests` will always be the same on all shards, right? That's a hard requirement for bandwidth scheduler, without it the state on shards will deviate.

I guess it's an argument against putting bandwidth requests in `CongestionInfo`...